OpenSim/Moco/MocoGoal/MocoImpulseTrackingGoal.cpp

```diff
@@ -0,0 +1,362 @@
+        OPENSIM_THROW_FRMOBJ(Exception,
+            "Expected a single (1) contact group, but "
+            "got '{}'.",
+            getProperty_contact_groups().size());
+    
+    setRequirements(1, 1, SimTK::Stage::Velocity);
+}
+
+int MocoImpulseTrackingGoal::findRecordOffset(
+    const MocoImpulseTrackingGoalGroup& group,
+    const SmoothSphereHalfSpaceForce& contactForce,
+    const std::string& appliedToBody) const {
+
+    // Is the ExternalForce applied to the sphere's body?
+    const auto& sphereBase =
+        contactForce.getConnectee<ContactSphere>("sphere")
+        .getConnectee<PhysicalFrame>("frame")
+        .findBaseFrame();
+    const std::string& sphereBaseName = sphereBase.getName();
+    if (sphereBaseName == appliedToBody) {
+        // We want the first 3 entries in
+        // SmoothSphereHalfSpaceForce::getRecordValues(), which contain
+        // forces on the sphere.
+        return 0;
+    }
+
+    // Is the ExternalForce applied to the half space's body?
+    const auto& halfSpaceBase =
+        contactForce.getConnectee<ContactHalfSpace>("half_space")
+        .getConnectee<PhysicalFrame>("frame")
+        .findBaseFrame();
+    const std::string& halfSpaceBaseName = halfSpaceBase.getName();
+    if (halfSpaceBaseName == appliedToBody) {
+        // We want the forces applied to the half space, which are
+        // entries 6, 7, 8 in
+        // SmoothSphereHalfSpaceForce::getRecordValues().
+        return 6;
+    }
+
+    // Check the group's alternative frames.
+    // Check each alternative frame path in the order provided. As soon as one
+    // of these paths matches the name of the sphere base frame or
+    // half space base frame, use the contact forces applied to that base frame.
+    const auto& sphereBasePath = sphereBase.getAbsolutePathString();
+    const auto& halfSpaceBasePath = halfSpaceBase.getAbsolutePathString();
+    for (int ia = 0; ia < group.getProperty_alternative_frame_paths().size();
+        ++ia) {
+        const auto& path = group.get_alternative_frame_paths(ia);
+        if (path == sphereBasePath) { return 0; }
+        if (path == halfSpaceBasePath) { return 6; }
+    }
+
+    OPENSIM_THROW_FRMOBJ(Exception,
+        "Contact force '{}' has sphere base frame '{}' and half space base "
+        "frame '{}'. One of these frames should match the applied_to_body "
+        "setting ('{}') of ExternalForce '{}', or match one of the "
+        "alternative_frame_paths, but no match found.",
+        contactForce.getAbsolutePathString(), sphereBaseName,
+        halfSpaceBaseName, appliedToBody, group.get_external_force_name());
+}
+
+void MocoImpulseTrackingGoal::calcIntegrandImpl(
+    const IntegrandInput& input, double& integrand) const {
+    const auto& state = input.state;
+    const auto& time = state.getTime();
+    getModel().realizeVelocity(state);
+    SimTK::Vector timeVec(1, time);
+
+    integrand = 0;
+    SimTK::Vec3 force_ref;
```

Comment:
Since we're only storing one spline for the impulse, all the `Vec3` quantities in this method can now just be `double`s.